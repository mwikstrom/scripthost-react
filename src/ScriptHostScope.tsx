import React, { createContext, FC, ReactNode, useContext, useMemo } from "react";
import { ScriptHost, ScriptHostOptions } from "scripthost";
import { BrowserSandbox } from "scripthost-browser";

/**
 * @public
 */
export interface ScriptHostScopeProps {
    children?: ReactNode;
}

/**
 * @public
 */
export const ScriptHostScope: FC<ScriptHostScopeProps> = props => {
    const {  children } = props;
    const host = useMemo(() => new ScriptHost(defaultHostOptions), []);
    return (
        <ScriptHostContext.Provider
            value={host}
            children={children}
        />
    );
};

/**
 * @public
 */
export const useScriptHost = (): ScriptHost => useContext(ScriptHostContext);

const createBrowserSandbox = () => new BrowserSandbox();

const defaultHostOptions: ScriptHostOptions = { createSandbox: createBrowserSandbox };

const ScriptHostContext = createContext<ScriptHost>(new ScriptHost(defaultHostOptions));
