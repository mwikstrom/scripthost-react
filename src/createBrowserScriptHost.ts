import { ScriptHost, ScriptHostOptions } from "scripthost";
import { BrowserSandbox, BrowserSandboxOptions } from "scripthost-browser";

/**
 * @public
 */
export function createBrowserScriptHost(
    options: ScriptHostOptions & Pick<BrowserSandboxOptions, "unsafe"> = {}
): ScriptHost {
    const { unsafe, ...hostOptions } = options;
    return new ScriptHost(() => new BrowserSandbox({ unsafe }), { ...hostOptions });
}
