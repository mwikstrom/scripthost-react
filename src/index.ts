/**
 * React bindings for Scripthost
 * @packageDocumentation
 */

export * from "./ScriptHostScope";
export * from "./useObservedScript";
export * from "./useScriptInvoker";
