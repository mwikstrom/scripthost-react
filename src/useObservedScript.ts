import { useEffect, useState } from "react";
import { ScriptEvalOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export interface ObservedScript {
    result: ScriptValue;
    ready: boolean;
    failed: boolean;
    error: unknown;
}

/**
 * @public
 */
export type UseObservedScriptOptions = Pick<ScriptEvalOptions, "instanceId" | "timeout">;

/**
 * @public
 */
export function useObservedScript(script: string, options: UseObservedScriptOptions = {}): ObservedScript {
    const { instanceId, timeout } = options;
    const host = useScriptHost();
    const [output, setOutput] = useState(initialOutput);

    useEffect(() => {
        const onNext = (result: ScriptValue) => setOutput(successOutput(result));
        const onError = (error: unknown) => setOutput(errorOutput(error));
        setOutput(initialOutput);
        return host.observe(script, { instanceId, timeout, onNext, onError });
    }, [host, script, instanceId, timeout, setOutput]);

    return output;
}

const initialOutput: ObservedScript = Object.freeze({
    result: undefined,
    ready: false,
    failed: false,
    error: undefined,
});

const successOutput = (result: ScriptValue): ObservedScript => Object.freeze({
    result,
    ready: true,
    failed: false,
    error: undefined,
});

const errorOutput = (error: unknown): ObservedScript => Object.freeze({
    result: undefined,
    ready: true,
    failed: true,
    error,
});
