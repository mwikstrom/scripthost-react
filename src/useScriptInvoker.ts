import { useCallback } from "react";
import { ScriptEvalOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export type ScriptInvoker = () => Promise<ScriptValue>;

/**
 * @public
 */
export type UseScriptInvokerOptions = Pick<ScriptEvalOptions, "instanceId" | "timeout">;

/**
 * @public
 */
export function useScriptInvoker(script: string, options: UseScriptInvokerOptions = {}): ScriptInvoker {
    const { instanceId, timeout } = options;
    const host = useScriptHost();
    return useCallback(() => host.eval(script, { instanceId, timeout }), [host, script, instanceId, timeout]);
}