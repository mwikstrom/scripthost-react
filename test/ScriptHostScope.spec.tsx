import React from "react";
import renderer from "react-test-renderer";
import { ScriptHostScope } from "../src";

describe("ScriptHostScope", () => {
    it("can render without props", () => {
        const component = renderer.create(<ScriptHostScope/>);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});