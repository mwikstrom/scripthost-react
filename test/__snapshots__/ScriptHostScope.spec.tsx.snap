// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScriptHostScope can render without props 1`] = `null`;
